File: harz-edge-telecom/edge-rebuild-v2.2/core/src/main/kotlin/com/harz/edgetelecom/mesh/MeshCore.kt
package com.harz.edgetelecom.mesh

/**
 * MeshCore — the HARZ-Mesh data plane, radio-agnostic, JVM-pure.
 *
 * Implements the frozen protocol EXACTLY (the acceptance harness's 8 checks):
 *  T1  direct 2-node delivery (HELLO -> DATA -> ACK)
 *  T2  multi-hop A->B->C: relay forwarding, route append [A,B], hops=1, ACK return
 *  T2b payload integrity: chunked reassembly stays byte-identical
 *  T3  TTL attack (G2-C): ttl 0 never forwarded, ttl 1 stops at relay, ttl 2+ traverses
 *  T4  dedup: same (src,seq) delivered exactly once
 *  T5  outage recovery (G2-E): peer loss -> frames held -> byte-identical flush on return
 *
 * FROZEN RULE (Sep 6 report §3): NO retransmit in the app data plane.
 * Loss is counted honestly; the harness's retransmit was harness-only.
 */
class MeshCore(
    val self: EdgeId,
    private val listener: Listener
) {
    interface Listener {
        fun onData(src: EdgeId, seq: Long, hops: Int, route: List<EdgeId>, payload: ByteArray)
        fun onAck(src: EdgeId, seq: Long)
        fun onPeerJoined(peer: EdgeId, via: String)
        fun onPeerLost(peer: EdgeId)
        fun onMetric(event: String, detail: String)
    }

    /** Registered peers: edgeId -> transport. */
    private val peers = LinkedHashMap<EdgeId, Transport>()
    /** transport -> registered edgeId (reverse index). */
    private val transportIds = HashMap<Transport, EdgeId>()
    private val dedup = DedupCache()
    val store = StoreAndForward()
    var seqCounter: Long = 0
        private set

    // ---------------------------------------------------------------- wiring

    /** A new transport appeared (GO accepted / client socket connected). Await HELLO. */
    fun attach(t: Transport) {
        t.onFrame { raw -> safeHandle(raw, t) }
    }

    /** Transport closed (peer died / socket reset) — G2-E observation point. */
    fun detach(t: Transport) {
        val id = transportIds.remove(t) ?: return
        peers.remove(id)
        listener.onPeerLost(id)
        listener.onMetric("peer_lost", id.value)
    }

    private fun safeHandle(raw: ByteArray, t: Transport) {
        val f = try { MeshFrame.decode(raw) } catch (e: Exception) {
            listener.onMetric("bad_frame", e.message ?: "malformed")
            return
        }
        handle(f, t)
    }

    private fun handle(f: MeshFrame, from: Transport) {
        when (f.type) {
            MeshProtocol.TYPE_HELLO -> {
                register(f.src, from)
                from.send(MeshFrame.helloAck(self, f.src, f.seq).encode())
            }
            MeshProtocol.TYPE_HELLO_ACK -> register(f.src, from)
            MeshProtocol.TYPE_BYE -> {
                transportIds.remove(from)?.let { peers.remove(it); listener.onPeerLost(it) }
            }
            MeshProtocol.TYPE_DATA -> onDataFrame(f, from)
            MeshProtocol.TYPE_ACK -> onAckFrame(f, from)
        }
    }

    private fun register(id: EdgeId, t: Transport) {
        val isNew = !peers.containsKey(id)
        transportIds[t] = id
        peers[id] = t
        if (isNew) {
            listener.onPeerJoined(id, t.describe())
            listener.onMetric("peer_joined", id.value)
        } else {
            listener.onMetric("peer_back", id.value)
        }
        // Flush on EVERY successful registration: a returning peer must receive
        // held frames (T5 shape). Receiver dedup guards against double-flush.
        flushStore(t, id)
    }

    /** On (re)connection: drain held frames toward the returning peer. Byte-identical. */
    private fun flushStore(t: Transport, peerId: EdgeId) {
        val held = store.drain(peerId)
        if (held.isNotEmpty()) listener.onMetric("flush", "${peerId.value}:${held.size}")
        for (frame in held) t.send(frame.encode())
        // Frames held for OTHER peers: the returning peer may be the relay toward them (T5 shape).
        val others = ArrayList<MeshFrame>()
        for (entry in store.heldSnapshot()) {
            if (entry.key != peerId.value) { others.addAll(store.drainKey(entry.key)); }
        }
        for (frame in others) t.send(frame.encode())
    }

    // ---------------------------------------------------------------- DATA plane

    private fun onDataFrame(f: MeshFrame, from: Transport) {
        if (f.dst.value == self.value) {
            // Deliver locally — exactly once (T4).
            if (dedup.seen(f.src, f.seq)) { listener.onMetric("dup_dropped", "${f.src.value}:${f.seq}"); return }
            listener.onData(f.src, f.seq, f.hops, f.route, f.payload)
            // ACK return path along the way it came (T2 check 4).
            from.send(MeshFrame.ack(self, f.src, f.seq, f.route).encode())
            return
        }
        // Not for me: relay plane. Dedup before forwarding (T4 at the relay).
        if (dedup.seen(f.src, f.seq)) { listener.onMetric("dup_dropped", "${f.src.value}:${f.seq}"); return }

        // G2-C frozen TTL semantics:
        if (f.ttl <= MeshProtocol.Ttl.STOP_AT_RELAY) {
            // ttl 0: never forwarded. ttl 1: stops HERE at the relay.
            listener.onMetric("ttl_stop", "ttl=${f.ttl} seq=${f.seq} at=${self.value}")
            return
        }

        val forwarded = f.copy(ttl = f.ttl - 1, hops = f.hops + 1, route = f.route + self)
        val direct = peers[f.dst]
        if (direct != null && direct.isConnected()) {
            direct.send(forwarded.encode())
            listener.onMetric("forwarded", "${f.src.value}->${f.dst.value} via ${self.value}")
        } else {
            // Flood toward other peers (relay shape: GO with several clients).
            var sentTo = 0
            for ((id, t) in peers) {
                if (id == f.src) continue // never bounce back to sender
                if (id == self) continue
                if (t.isConnected()) { t.send(forwarded.encode()); sentTo++ }
            }
            if (sentTo > 0) {
                listener.onMetric("forwarded", "${f.src.value}->${f.dst.value} flood x$sentTo")
            } else {
                store.hold(f.dst, f.copy()) // hold ORIGINAL frame; flush is byte-identical
                listener.onMetric("held", "${f.dst.value}:${f.seq}")
            }
        }
    }

    private fun onAckFrame(f: MeshFrame, from: Transport) {
        if (f.dst.value == self.value) {
            listener.onAck(f.src, f.seq)
            return
        }
        // Relay the ACK back toward its destination the way it came.
        val direct = peers[f.dst]
        if (direct != null && direct.isConnected()) {
            direct.send(f.encode())
        } else {
            for ((id, t) in peers) {
                if (id == f.src) continue
                if (t.isConnected()) { t.send(f.encode()); return }
            }
            store.hold(f.dst, f.copy())
            listener.onMetric("held_ack", "${f.dst.value}:${f.seq}")
        }
    }

    // ---------------------------------------------------------------- send API

    /**
     * Send one numbered chunk to dst. Route starts with the sender ([A] in the
     * desk's check-3 shape); each forwarder appends itself (-> [A,B]).
     * Pacing (20 ms) is the caller's loop — the core never blocks the radio thread.
     */
    fun sendChunk(dst: EdgeId, chunk: ByteArray, ttl: Int = MeshProtocol.DEFAULT_TTL): Long {
        val seq = seqCounter++
        val frame = MeshFrame(
            type = MeshProtocol.TYPE_DATA, src = self, dst = dst, seq = seq,
            ttl = ttl, hops = 0, route = listOf(self), payload = chunk
        )
        val direct = peers[dst]
        if (direct != null && direct.isConnected()) {
            direct.send(frame.encode())
        } else {
            // No direct link: send toward any relay peer, else hold (T5).
            var relayed = false
            for ((id, t) in peers) {
                if (t.isConnected()) { t.send(frame.encode()); relayed = true; break }
            }
            if (!relayed) { store.hold(dst, frame); listener.onMetric("held", "${dst.value}:${seq}") }
        }
        return seq
    }

    /** Split a payload into protocol-sized chunks (900 B). Caller paces and numbers. */
    fun chunkPayload(data: ByteArray, chunkSize: Int = MeshProtocol.CHUNK_SIZE): List<ByteArray> =
        if (data.isEmpty()) listOf(ByteArray(0))
        else data.toList().chunked(chunkSize).map { it.toByteArray() }

    fun peersOnline(): List<EdgeId> = peers.keys.toList()
}

File: harz-reach/v0.3-mesh/mesh-core.js
// HARZ MESH CORE v2.1 — JS twin of the frozen Kotlin MeshCore.kt data plane.
// Same frozen rules, unchanged: HELLO registration + HELLO_ACK, dedup on (src,seq),
// G2-C TTL semantics (0 never forwarded, 1 stops at relay, 2+ traverses),
// hop counting + route append, ACK return along the route, store-and-forward
// (G2-E: hold on peer loss, byte-identical flush on return, loss counted honestly).
// Radio-agnostic Transport interface: { send(bytes), isConnected(), close() }.

import * as F from "./mesh-frame.js";

export class DedupCache {
  constructor(capacity = 4096) { this.cap = capacity; this.map = new Map(); }
  seen(src, seq) { const k = src + ":" + seq; if (this.map.has(k)) return true; this.map.set(k, 1); if (this.map.size > this.cap) this.map.delete(this.map.keys().next().value); return false; }
  size() { return this.map.size; }
}

export class StoreAndForward {
  constructor(maxHeld = 1024) { this.maxHeld = maxHeld; this.held = new Map(); }
  hold(dst, frame) { if (this.totalHeld() >= this.maxHeld) return; if (!this.held.has(dst)) this.held.set(dst, []); this.held.get(dst).push(frame); }
  drain(dst) { return this.held.delete(dst) ? arguments[2] || [] : []; }
}

export class MeshNode {
  constructor(id, listener) {
    this.self = id;
    this.listener = listener || {};
    this.peers = new Map();       // edgeId -> transport
    this.transports = new Map();  // transport -> edgeId
    this.dedup = new DedupCache();
    this.store = new Map();       // dst -> frames held
    this.seqCounter = 0;
    this.metrics = [];
  }

  onMetric(ev, detail) { this.metrics.push(ev + " " + detail); if (this.listener.onMetric) this.listener.onMetric(ev, detail); }

  attach(t) {
    t.onFrame = raw => {
      let f;
      try { f = F.decodeFrame(raw); } catch (e) { this.onMetric("bad_frame", e.message); return; }
      this.handle(f, t);
    };
  }

  detach(t) {
    const id = this.transports.get(t);
    if (!id) return;
    this.transports.delete(t); this.peers.delete(id);
    if (this.listener.onPeerLost) this.listener.onPeerLost(id);
    this.onMetric("peer_lost", id);
  }

  handle(f, t) {
    switch (f.type) {
      case F.TYPE_HELLO:
        this.register(f.src, t);
        t.send(F.encodeFrame(F.helloAck(this.self, f.src, f.seq)));
        break;
      case F.TYPE_HELLO_ACK: this.register(f.src, t); break;
      case F.TYPE_BYE: {
        const id = this.transports.get(t);
        if (id) { this.transports.delete(t); this.peers.delete(id); if (this.listener.onPeerLost) this.listener.onPeerLost(id); }
        break;
      }
      case F.TYPE_DATA: this.onDataFrame(f, t); break;
      case F.TYPE_ACK: this.onAckFrame(f, t); break;
    }
  }

  register(id, t) {
    const isNew = !this.peers.has(id);
    this.transports.set(t, id);
    this.peers.set(id, t);
    if (isNew) { if (this.listener.onPeerJoined) this.listener.onPeerJoined(id); this.onMetric("peer_joined", id); }
    else this.onMetric("peer_back", id);
    this.flushStore(t, id);
  }

  flushStore(t, peerId) {
    const held = this.store.get(peerId);
    if (held && held.length) { this.onMetric("flush", peerId + ":" + held.length); for (const fr of held) t.send(F.encodeFrame(fr)); this.store.delete(peerId); }
    // frames held for others: the returning peer may relay toward them (T5 shape)
    const others = [];
    for (const [k, list] of this.store) if (k !== peerId) { others.push(...list); this.store.delete(k); }
    for (const fr of others) t.send(F.encodeFrame(fr));
  }

  onDataFrame(f, from) {
    if (f.dst === this.self) {
      if (this.dedup.seen(f.src, f.seq)) { this.onMetric("dup_dropped", f.src + ":" + f.seq); return; } // T4
      if (this.listener.onData) this.listener.onData(f.src, f.seq, f.hops, f.route, f.payload);
      from.send(F.encodeFrame(F.ack(this.self, f.src, f.seq, f.route))); // T2 ACK return
      return;
    }
    if (this.dedup.seen(f.src, f.seq)) { this.onMetric("dup_dropped", f.src + ":" + f.seq); return; }
    if (f.ttl <= 1) { this.onMetric("ttl_stop", "ttl=" + f.ttl + " seq=" + f.seq + " at=" + this.self); return; } // G2-C
    const forwarded = { ...f, ttl: f.ttl - 1, hops: f.hops + 1, route: [...f.route, this.self] };
    const direct = this.peers.get(f.dst);
    if (direct && direct.isConnected()) { direct.send(F.encodeFrame(forwarded)); this.onMetric("forwarded", f.src + "->" + f.dst + " via " + this.self); }
    else {
      let sentTo = 0;
      for (const [id, tp] of this.peers) {
        if (id === f.src || id === this.self) continue;
        if (tp.isConnected()) { tp.send(F.encodeFrame(forwarded)); sentTo++; }
      }
      if (sentTo > 0) this.onMetric("forwarded", f.src + "->" + f.dst + " flood x" + sentTo);
      else { if (!this.store.has(f.dst)) this.store.set(f.dst, []); if (this.totalHeld() < 1024) this.store.get(f.dst).push(f); this.onMetric("held", f.dst + ":" + f.seq); }
    }
  }

  onAckFrame(f, from) {
    if (f.dst === this.self) { if (this.listener.onAck) this.listener.onAck(f.src, f.seq); return; }
    const direct = this.peers.get(f.dst);
    if (direct && direct.isConnected()) direct.send(F.encodeFrame(f));
    else {
      for (const [id, tp] of this.peers) {
        if (id === f.src) continue;
        if (tp.isConnected()) { tp.send(F.encodeFrame(f)); return; }
      }
      if (!this.store.has(f.dst)) this.store.set(f.dst, []);
      if (this.totalHeld() < 1024) this.store.get(f.dst).push(f);
      this.onMetric("held_ack", f.dst + ":" + f.seq);
    }
  }

  totalHeld() { let n = 0; for (const l of this.store.values()) n += l.length; return n; }
  heldFor(dst) { return (this.store.get(dst) || []).length; }
  peersOnline() { return [...this.peers.keys()]; }

  chunkPayload(buf, size = F.CHUNK_SIZE) {
    const out = [];
    for (let i = 0; i < buf.length; i += size) out.push(buf.subarray(i, Math.min(i + size, buf.length)));
    return out.length ? out : [Buffer.alloc(0)];
  }

  sendChunk(dst, chunk, ttl = F.DEFAULT_TTL) {
    const seq = this.seqCounter++;
    const frame = { type: F.TYPE_DATA, src: this.self, dst, seq, ttl, hops: 0, route: [this.self], payload: chunk };
    const direct = this.peers.get(dst);
    if (direct && direct.isConnected()) { direct.send(F.encodeFrame(frame)); return seq; }
    // not directly connected: hand to any connected peer (relay/flood shape)
    for (const [id, tp] of this.peers) {
      if (id === this.self) continue;
      if (tp.isConnected()) { tp.send(F.encodeFrame(frame)); return seq; }
    }
    if (!this.store.has(dst)) this.store.set(dst, []);
    if (this.totalHeld() < 1024) this.store.get(dst).push(frame); // hold ORIGINAL frame
    this.onMetric("held", dst + ":" + seq);
    return seq;
  }
}
